src/System.Net.Security/src/System/Net/_SecureChannel.cs

```diff
@@ -0,0 +1,1487 @@
+                try
+                {
+                    X509Certificate2Collection dummyCollection;
+                    remoteCert = GetRemoteCertificate(out dummyCollection);
+                    clientCertificate = _certSelectionDelegate(_hostName, ClientCertificates, remoteCert, issuers);
+                }
+                finally
+                {
+                    if (remoteCert != null)
+                        remoteCert.Dispose();
+                }
+
+
+                if (clientCertificate != null)
+                {
+                    if (_credentialsHandle == null)
+                        sessionRestartAttempt = true;
+                    filteredCerts.Add(clientCertificate);
+                    if (Logging.On) Logging.PrintInfo(Logging.Web, this, SR.net_log_got_certificate_from_delegate);
+                }
+                else
+                {
+                    if (ClientCertificates.Count == 0)
+                    {
+                        if (Logging.On) Logging.PrintInfo(Logging.Web, this, SR.net_log_no_delegate_and_have_no_client_cert);
+                        sessionRestartAttempt = true;
+                    }
+                    else
+                    {
+                        if (Logging.On) Logging.PrintInfo(Logging.Web, this, SR.net_log_no_delegate_but_have_client_cert);
+                    }
+                }
+            }
+            else if (_credentialsHandle == null && _clientCertificates != null && _clientCertificates.Count > 0)
+            {
+                // This is where we attempt to restart a session by picking the FIRST cert from the collection.
+                // Otherwise it is either server sending a client cert request or the session is renegotiated.
+                clientCertificate = ClientCertificates[0];
+                sessionRestartAttempt = true;
+                if (clientCertificate != null)
+                    filteredCerts.Add(clientCertificate);
+                if (Logging.On) Logging.PrintInfo(Logging.Web, this, SR.Format(SR.net_log_attempting_restart_using_cert, (clientCertificate == null ? "null" : clientCertificate.ToString(true))));
+            }
+            else if (_clientCertificates != null && _clientCertificates.Count > 0)
+            {
+                //
+                // This should be a server request for the client cert sent over currently anonymous sessions.
+                //
+                issuers = GetIssuers();
+
+                if (Logging.On)
+                {
+                    if (issuers == null || issuers.Length == 0)
+                        Logging.PrintInfo(Logging.Web, this, SR.net_log_no_issuers_try_all_certs);
+                    else
+                        Logging.PrintInfo(Logging.Web, this, SR.Format(SR.net_log_server_issuers_look_for_matching_certs, issuers.Length));
+                }
+
+                for (int i = 0; i < _clientCertificates.Count; ++i)
```

Comment:
And... now I see that I mis-aligned the break when reading the code; so it's really FindCandidate<b>s</b>.  Having that inner loop be a method that returns true/false seems very good to me.